README.md
[![view on npm](https://img.shields.io/npm/v/@uttori/plugin-generator-sitemap.svg)](https://www.npmjs.com/package/@uttori/plugin-generator-sitemap)
[![npm module downloads](https://img.shields.io/npm/dt/@uttori/plugin-generator-sitemap)](https://www.npmjs.com/package/@uttori/plugin-generator-sitemap)
[![Build Status](https://travis-ci.com/uttori/uttori-plugin-generator-sitemap.svg?branch=master)](https://travis-ci.com/uttori/uttori-plugin-generator-sitemap)
[![Dependency Status](https://david-dm.org/uttori/uttori-plugin-generator-sitemap.svg)](https://david-dm.org/uttori/uttori-plugin-generator-sitemap)
[![Coverage Status](https://coveralls.io/repos/uttori/uttori-plugin-generator-sitemap/badge.svg?branch=master)](https://coveralls.io/r/uttori/uttori-plugin-generator-sitemap?branch=master)
[![Tree-Shaking Support](https://badgen.net/bundlephobia/tree-shaking/@uttori/plugin-generator-sitemap)](https://bundlephobia.com/result?p=@uttori/plugin-generator-sitemap)
[![Dependency Count](https://badgen.net/bundlephobia/dependency-count/@uttori/plugin-generator-sitemap)](https://bundlephobia.com/result?p=@uttori/plugin-generator-sitemap)
[![Minified + GZip](https://badgen.net/bundlephobia/minzip/@uttori/plugin-generator-sitemap)](https://bundlephobia.com/result?p=@uttori/plugin-generator-sitemap)
[![Minified](https://badgen.net/bundlephobia/min/@uttori/plugin-generator-sitemap)](https://bundlephobia.com/result?p=@uttori/plugin-generator-sitemap)

# Uttori Sitemap Generator

A plugin to generating a XML sitemap.

## Install

```bash
npm install --save @uttori/plugin-generator-sitemap
```

## Config

```js
{
  // Registration Events
  events: {
    callback: ['document-save', 'document-delete'],
    validateConfig: ['validate-config'],
  },

  [SitemapGenerator.configKey]: {
    ...
  }
}
```

* * *

## API Reference

## Classes

<dl>
<dt><a href="#SitemapGenerator">SitemapGenerator</a></dt>
<dd><p>Uttori Sitemap Generator</p>
<p>Generates a valid sitemap.xml file for submitting to search engines.</p>
</dd>
</dl>

## Functions

<dl>
<dt><a href="#debug">debug()</a> : <code>function</code></dt>
<dd></dd>
</dl>

## Typedefs

<dl>
<dt><a href="#SitemapGeneratorConfig">SitemapGeneratorConfig</a> : <code>object</code></dt>
<dd></dd>
</dl>

<a name="SitemapGenerator"></a>

## SitemapGenerator
Uttori Sitemap Generator

Generates a valid sitemap.xml file for submitting to search engines.

**Kind**: global class  

* [SitemapGenerator](#SitemapGenerator)
    * [.configKey](#SitemapGenerator.configKey) ⇒ <code>string</code>
    * [.defaultConfig()](#SitemapGenerator.defaultConfig) ⇒ [<code>SitemapGeneratorConfig</code>](#SitemapGeneratorConfig)
    * [.validateConfig(config, [_context])](#SitemapGenerator.validateConfig)
    * [.register(context)](#SitemapGenerator.register)
    * [.callback(_document, context)](#SitemapGenerator.callback) ⇒ <code>Promise</code>
    * [.generateSitemap(context)](#SitemapGenerator.generateSitemap) ⇒ <code>Promise</code>

<a name="SitemapGenerator.configKey"></a>

### SitemapGenerator.configKey ⇒ <code>string</code>
The configuration key for plugin to look for in the provided configuration.

**Kind**: static property of [<code>SitemapGenerator</code>](#SitemapGenerator)  
**Returns**: <code>string</code> - The configuration key.  
**Example** *(SitemapGenerator.configKey)*  
```js
const config = { ...SitemapGenerator.defaultConfig(), ...context.config[SitemapGenerator.configKey] };
```
<a name="SitemapGenerator.defaultConfig"></a>

### SitemapGenerator.defaultConfig() ⇒ [<code>SitemapGeneratorConfig</code>](#SitemapGeneratorConfig)
The default configuration.

**Kind**: static method of [<code>SitemapGenerator</code>](#SitemapGenerator)  
**Returns**: [<code>SitemapGeneratorConfig</code>](#SitemapGeneratorConfig) - The configuration.  
**Example** *(SitemapGenerator.defaultConfig())*  
```js
const config = { ...SitemapGenerator.defaultConfig(), ...context.config[SitemapGenerator.configKey] };
```
<a name="SitemapGenerator.validateConfig"></a>

### SitemapGenerator.validateConfig(config, [_context])
Validates the provided configuration for required entries.

**Kind**: static method of [<code>SitemapGenerator</code>](#SitemapGenerator)  

| Param | Type | Description |
| --- | --- | --- |
| config | <code>object</code> | A configuration object. |
| config.configKey | [<code>SitemapGeneratorConfig</code>](#SitemapGeneratorConfig) | A configuration object specifically for this plugin. |
| [_context] | <code>object</code> | A Uttori-like context (unused). |

**Example** *(SitemapGenerator.validateConfig(config, _context))*  
```js
SitemapGenerator.validateConfig({ ... });
```
<a name="SitemapGenerator.register"></a>

### SitemapGenerator.register(context)
Register the plugin with a provided set of events on a provided Hook system.

**Kind**: static method of [<code>SitemapGenerator</code>](#SitemapGenerator)  

| Param | Type | Description |
| --- | --- | --- |
| context | <code>object</code> | A Uttori-like context. |
| context.hooks | <code>object</code> | An event system / hook system to use. |
| context.hooks.on | <code>function</code> | An event registration function. |
| context.config | <code>object</code> | A provided configuration to use. |
| context.config.events | <code>object</code> | An object whose keys correspong to methods, and contents are events to listen for. |

**Example** *(SitemapGenerator.register(context))*  
```js
const context = {
  hooks: {
    on: (event, callback) => { ... },
  },
  config: {
    [SitemapGenerator.configKey]: {
      ...,
      events: {
        callback: ['document-save', 'document-delete'],
        validateConfig: ['validate-config'],
      },
    },
  },
};
SitemapGenerator.register(context);
```
<a name="SitemapGenerator.callback"></a>

### SitemapGenerator.callback(_document, context) ⇒ <code>Promise</code>
Wrapper function for calling generating and writing the sitemap file.

**Kind**: static method of [<code>SitemapGenerator</code>](#SitemapGenerator)  
**Returns**: <code>Promise</code> - The provided document.  

| Param | Type | Description |
| --- | --- | --- |
| _document | <code>object</code> | A Uttori document (unused). |
| context | <code>object</code> | A Uttori-like context. |
| context.config | <code>object</code> | A provided configuration to use. |
| context.config.configKey | [<code>SitemapGeneratorConfig</code>](#SitemapGeneratorConfig) | A configuration object specifically for this plugin. |
| context.hooks | <code>object</code> | An event system / hook system to use. |
| context.hooks.on | <code>function</code> | An event registration function. |
| context.hooks.fetch | <code>function</code> | An event dispatch function that returns an array of results. |

**Example** *(SitemapGenerator.callback(_document, context))*  
```js
const context = {
  config: {
    [SitemapGenerator.configKey]: {
      ...,
    },
  },
  hooks: {
    on: (event) => { ... }
  },
};
SitemapGenerator.callback(null, context);
```
<a name="SitemapGenerator.generateSitemap"></a>

### SitemapGenerator.generateSitemap(context) ⇒ <code>Promise</code>
Generates a sitemap from the provided context.

**Kind**: static method of [<code>SitemapGenerator</code>](#SitemapGenerator)  
**Returns**: <code>Promise</code> - The generated sitemap.  

| Param | Type | Description |
| --- | --- | --- |
| context | <code>object</code> | A Uttori-like context. |
| context.config | <code>object</code> | A provided configuration to use. |
| context.config.configKey | [<code>SitemapGeneratorConfig</code>](#SitemapGeneratorConfig) | A configuration object specifically for this plugin. |
| context.config.events | <code>object</code> | An object whose keys correspong to methods, and contents are events to listen for. |
| context.hooks | <code>object</code> | An event system / hook system to use. |
| context.hooks.on | <code>function</code> | An event registration function. |
| context.hooks.fetch | <code>function</code> | An event dispatch function that returns an array of results. |

**Example** *(SitemapGenerator.callback(_document, context))*  
```js
const context = {
  config: {
    [SitemapGenerator.configKey]: {
      ...,
    },
  },
  hooks: {
    on: (event) => { ... },
    fetch: (event, query) => { ... },
  },
};
SitemapGenerator.generateSitemap(context);
```
<a name="debug"></a>

## debug() : <code>function</code>
**Kind**: global function  
<a name="SitemapGeneratorConfig"></a>

## SitemapGeneratorConfig : <code>object</code>
**Kind**: global typedef  
**Properties**

| Name | Type | Default | Description |
| --- | --- | --- | --- |
| urls | <code>Array.&lt;object&gt;</code> |  | A collection of Uttori documents. |
| [url_filters] | <code>Array.&lt;RegExp&gt;</code> |  | A collection of Regular Expression URL filters to exclude documents. |
| base_url | <code>string</code> |  | The base URL (ie https://domain.tld) for all documents. |
| directory | <code>string</code> |  | The path to the location you want the sitemap file to be written to. |
| [filename] | <code>string</code> | <code>&quot;&#x27;sitemap&#x27;&quot;</code> | The file name to use for the generated file. |
| [extension] | <code>string</code> | <code>&quot;&#x27;xml&#x27;&quot;</code> | The file extension to use for the generated file. |
| [page_priority] | <code>string</code> | <code>&quot;&#x27;0.08&#x27;&quot;</code> | Sitemap default page priority. |
| [xml_header] | <code>string</code> |  | Sitemap XML Header, standard XML sitemap header is the default. |
| [xml_footer] | <code>string</code> |  | Sitemap XML Footer, standard XML sitemap closing tag is the default. |


* * *

## Tests

To run the test suite, first install the dependencies, then run `npm test`:

```bash
npm install
npm test
DEBUG=Uttori* npm test
```

## Contributors

* [Matthew Callis](https://github.com/MatthewCallis)

## License

* [MIT](LICENSE)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [4.0.0](https://github.com/uttori/uttori-plugin-generator-sitemap/compare/v3.1.4...v4.0.0) - 2020-01-23

- 🧰 Add ESM Support
- 🧰 Add explicit exports
- 🎁 Update dev dependencies

## [3.1.4](https://github.com/uttori/uttori-plugin-generator-sitemap/compare/v3.1.3...v3.1.4) - 2020-11-15

- 🎁 Update dev dependencies
- 🧰 Make `debug` an optional dependency

## [3.1.3](https://github.com/uttori/uttori-plugin-generator-sitemap/compare/v3.1.2...v3.1.3) - 2020-11-15

- 🎁 Update dev dependencies
- 🎁 Update README badge URLs
- 🧰 Change how types are made and rebuild types
- 🧰 Created this file


test/index.test.js
/* eslint-disable security/detect-non-literal-fs-filename */
const test = require('ava');
const { FileUtility } = require('uttori-utilities');
const SitemapGenerator = require('../src');

const config = SitemapGenerator.defaultConfig();
const context = {
  config: {
    [SitemapGenerator.configKey]: {
      ...config,
      base_url: 'https://domain.tld',
      directory: './',
    },
  },
  storageProvider: {
    getQuery: (_query) => [
      {
        updateDate: null,
        createDate: new Date('2019-04-20').toISOString(),
        slug: 'good-title',
      },
      {
        updateDate: new Date('2019-04-21').toISOString(),
        createDate: new Date('2019-04-21').toISOString(),
        slug: 'fake-title',
      },
    ],
  },
};

test('SitemapGenerator.register(context): can register', (t) => {
  t.notThrows(() => {
    SitemapGenerator.register({ hooks: { on: () => {} }, config: { [SitemapGenerator.configKey]: { events: { callback: [] } } } });
  });
});

test('SitemapGenerator.register(context): errors without event dispatcher', (t) => {
  t.throws(() => {
    SitemapGenerator.register({ hooks: {} });
  }, { message: 'Missing event dispatcher in \'context.hooks.on(event, callback)\' format.' });
});

test('SitemapGenerator.register(context): errors without events', (t) => {
  t.throws(() => {
    SitemapGenerator.register({ hooks: { on: () => {} }, config: { [SitemapGenerator.configKey]: { } } });
  }, { message: 'Missing events to listen to for in \'config.events\'.' });
});

test('SitemapGenerator.defaultConfig(): can return a default config', (t) => {
  t.notThrows(SitemapGenerator.defaultConfig);
});

test('SitemapGenerator.generateSitemap(_document, context): generates a valid sitemap', async (t) => {
  t.plan(1);
  const output = await SitemapGenerator.generateSitemap({ ...context, config: { ...context.config, [SitemapGenerator.configKey]: { ...context.config[SitemapGenerator.configKey], urls: [] } } });
  t.is(output, '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd"><url><loc>https://domain.tld/good-title</loc><lastmod>2019-04-20T00:00:00.000Z</lastmod><priority>0.80</priority></url><url><loc>https://domain.tld/fake-title</loc><lastmod>2019-04-21T00:00:00.000Z</lastmod><priority>0.80</priority></url></urlset>');
});

test('SitemapGenerator.generateSitemap(_document, context): generates a valid sitemap with no documents', async (t) => {
  t.plan(1);
  const output = await SitemapGenerator.generateSitemap({ storageProvider: { getQuery: () => [] }, config: { ...context.config, [SitemapGenerator.configKey]: { ...context.config[SitemapGenerator.configKey], urls: [] } } });
  t.is(output, '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd"></urlset>');
});

test('SitemapGenerator.generateSitemap(_document, context): filters out urls with a filter', async (t) => {
  t.plan(1);
  const output = await SitemapGenerator.generateSitemap({ ...context, config: { ...context.config, [SitemapGenerator.configKey]: { ...context.config[SitemapGenerator.configKey], urls: [], url_filters: [/fake-title$/i] } } });
  t.is(output, '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd"><url><loc>https://domain.tld/good-title</loc><lastmod>2019-04-20T00:00:00.000Z</lastmod><priority>0.80</priority></url></urlset>');
});

test('SitemapGenerator.validateConfig(config, _context): throws when sitemaps key is missing', (t) => {
  t.throws(() => {
    SitemapGenerator.validateConfig({});
  }, { message: 'sitemap configuration key is missing.' });
});

test('SitemapGenerator.validateConfig(config, _context): throws when urls is not an array', (t) => {
  t.throws(() => {
    SitemapGenerator.validateConfig({
      [SitemapGenerator.configKey]: {
        urls: {},
        url_filters: [],
      },
    });
  }, { message: 'urls should be an array of documents.' });
});

test('SitemapGenerator.validateConfig(config, _context): throws when url_filters is not an array', (t) => {
  t.throws(() => {
    SitemapGenerator.validateConfig({
      [SitemapGenerator.configKey]: {
        urls: [],
        url_filters: {},
      },
    });
  }, { message: 'url_filters should be an array of regular expression url filters.' });
});

test('SitemapGenerator.validateConfig(config, _context): throws when base_url is missing', (t) => {
  t.throws(() => {
    SitemapGenerator.validateConfig({
      [SitemapGenerator.configKey]: {
        urls: [],
        url_filters: [],
      },
    });
  }, { message: 'base_url is required should be an string of your base URL (ie https://domain.tld).' });
});

test('SitemapGenerator.validateConfig(config, _context): throws when base_url is not a string', (t) => {
  t.throws(() => {
    SitemapGenerator.validateConfig({
      [SitemapGenerator.configKey]: {
        urls: [],
        url_filters: [],
        base_url: null,
      },
    });
  }, { message: 'base_url is required should be an string of your base URL (ie https://domain.tld).' });
});

test('SitemapGenerator.validateConfig(config, _context): throws when directory is missing', (t) => {
  t.throws(() => {
    SitemapGenerator.validateConfig({
      [SitemapGenerator.configKey]: {
        urls: [],
        url_filters: [],
        base_url: 'https://domain.tld',
      },
    });
  }, { message: 'directory is required should be the path to the location you want the sitemap to be writtent to.' });
});

test('SitemapGenerator.validateConfig(config, _context): throws when directory is not a string', (t) => {
  t.throws(() => {
    SitemapGenerator.validateConfig({
      [SitemapGenerator.configKey]: {
        urls: [],
        url_filters: [],
        base_url: 'https://domain.tld',
        directory: null,
      },
    });
  }, { message: 'directory is required should be the path to the location you want the sitemap to be writtent to.' });
});

test('SitemapGenerator.validateConfig(config, _context): can validate', (t) => {
  t.notThrows(() => {
    SitemapGenerator.validateConfig({
      [SitemapGenerator.configKey]: {
        urls: [],
        url_filters: [],
        base_url: 'https://domain.tld',
        directory: './',
      },
    });
  });
});

test('SitemapGenerator.callback(_document, context): writes properly generated sitemap to desired location', async (t) => {
  await FileUtility.deleteFile('./', 'sitemap', 'xml');
  await SitemapGenerator.callback(null, { ...context, config: { ...context.config, [SitemapGenerator.configKey]: { ...context.config[SitemapGenerator.configKey], urls: [] } } });
  const output = await FileUtility.readFile('./', 'sitemap', 'xml');
  t.is(output, '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd"><url><loc>https://domain.tld/good-title</loc><lastmod>2019-04-20T00:00:00.000Z</lastmod><priority>0.80</priority></url><url><loc>https://domain.tld/fake-title</loc><lastmod>2019-04-21T00:00:00.000Z</lastmod><priority>0.80</priority></url></urlset>');
  await FileUtility.deleteFile('./', 'sitemap', 'xml');
});


types/index.d.ts
declare module "sitemap-generator" {
    export = SitemapGenerator;
    class SitemapGenerator {
        static get configKey(): string;
        static defaultConfig(): SitemapGeneratorConfig;
        static validateConfig(config: {
            configKey: SitemapGeneratorConfig;
        }, _context?: object): void;
        static register(context: {
            hooks: {
                on: Function;
            };
            config: {
                events: object;
            };
        }): void;
        static callback(_document: object, context: {
            config: {
                configKey: SitemapGeneratorConfig;
            };
            hooks: {
                on: Function;
                fetch: Function;
            };
        }): Promise<any>;
        static generateSitemap(context: {
            config: {
                configKey: SitemapGeneratorConfig;
                events: object;
            };
            hooks: {
                on: Function;
                fetch: Function;
            };
        }): Promise<any>;
    }
    namespace SitemapGenerator {
        export { SitemapGeneratorConfig };
    }
    type SitemapGeneratorConfig = {
        urls: object[];
        url_filters?: RegExp[];
        base_url: string;
        directory: string;
        filename?: string;
        extension?: string;
        page_priority?: string;
        xml_header?: string;
        xml_footer?: string;
    };
}
declare module "index" {
    export const SitemapGenerator: typeof import("sitemap-generator");
}


src/sitemap-generator.js
/** @type {Function} */
let debug = () => {}; try { debug = require('debug')('Uttori.Plugin.SitemapGenerator'); } catch {}
const fs = require('fs');
const sanitize = require('sanitize-filename');

/**
 * @typedef SitemapGeneratorConfig
 * @type {object}
 * @property {object[]} urls A collection of Uttori documents.
 * @property {RegExp[]} [url_filters] A collection of Regular Expression URL filters to exclude documents.
 * @property {string} base_url The base URL (ie https://domain.tld) for all documents.
 * @property {string} directory The path to the location you want the sitemap file to be written to.
 * @property {string} [filename='sitemap'] The file name to use for the generated file.
 * @property {string} [extension='xml'] The file extension to use for the generated file.
 * @property {string} [page_priority='0.08'] Sitemap default page priority.
 * @property {string} [xml_header] Sitemap XML Header, standard XML sitemap header is the default.
 * @property {string} [xml_footer] Sitemap XML Footer, standard XML sitemap closing tag is the default.
 */

/**
 * Uttori Sitemap Generator
 *
 * Generates a valid sitemap.xml file for submitting to search engines.
 *
 * @example <caption>SitemapGenerator</caption>
 * const sitemap = SitemapGenerator.generate({ ... });
 * @class
 */
class SitemapGenerator {
  /**
   * The configuration key for plugin to look for in the provided configuration.
   *
   * @static
   * @type {string}
   * @returns {string} The configuration key.
   * @example <caption>SitemapGenerator.configKey</caption>
   * const config = { ...SitemapGenerator.defaultConfig(), ...context.config[SitemapGenerator.configKey] };
   */
  static get configKey() {
    return 'uttori-plugin-generator-sitemap';
  }

  /**
   * The default configuration.
   *
   * @static
   * @returns {SitemapGeneratorConfig} The configuration.
   * @example <caption>SitemapGenerator.defaultConfig()</caption>
   * const config = { ...SitemapGenerator.defaultConfig(), ...context.config[SitemapGenerator.configKey] };
   */
  static defaultConfig() {
    return {
      urls: [],
      url_filters: [],
      base_url: '',
      directory: '',
      filename: 'sitemap',
      extension: 'xml',
      page_priority: '0.80',
      xml_header: '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">',
      xml_footer: '</urlset>',
    };
  }

  /**
   * Validates the provided configuration for required entries.
   *
   * @static
   * @param {object} config A configuration object.
   * @param {SitemapGeneratorConfig} config.configKey A configuration object specifically for this plugin.
   * @param {object} [_context] A Uttori-like context (unused).
   * @example <caption>SitemapGenerator.validateConfig(config, _context)</caption>
   * SitemapGenerator.validateConfig({ ... });
   */
  static validateConfig(config, _context) {
    debug('Validating config...');
    if (!config[SitemapGenerator.configKey]) {
      debug('Config Error: `sitemap` configuration key is missing.');
      throw new Error('sitemap configuration key is missing.');
    }
    if (!Array.isArray(config[SitemapGenerator.configKey].urls)) {
      debug('Config Error: `urls` should be an array of documents.');
      throw new Error('urls should be an array of documents.');
    }
    if (config[SitemapGenerator.configKey].url_filters && !Array.isArray(config[SitemapGenerator.configKey].url_filters)) {
      debug('Config Error: `url_filters` should be an array of regular expression url filters.');
      throw new Error('url_filters should be an array of regular expression url filters.');
    }
    if (!config[SitemapGenerator.configKey].base_url || typeof config[SitemapGenerator.configKey].base_url !== 'string') {
      debug('Config Error: `base_url` is required should be an string of your base URL (ie https://domain.tld).');
      throw new Error('base_url is required should be an string of your base URL (ie https://domain.tld).');
    }
    if (!config[SitemapGenerator.configKey].directory || typeof config[SitemapGenerator.configKey].directory !== 'string') {
      debug('Config Error: `directory` is required should be the path to the location you want the sitemap to be writtent to.');
      throw new Error('directory is required should be the path to the location you want the sitemap to be writtent to.');
    }
    debug('Validated config.');
  }

  /**
   * Register the plugin with a provided set of events on a provided Hook system.
   *
   * @param {object} context A Uttori-like context.
   * @param {object} context.hooks An event system / hook system to use.
   * @param {Function} context.hooks.on An event registration function.
   * @param {object} context.config A provided configuration to use.
   * @param {object} context.config.events An object whose keys correspong to methods, and contents are events to listen for.
   * @example <caption>SitemapGenerator.register(context)</caption>
   * const context = {
   *   hooks: {
   *     on: (event, callback) => { ... },
   *   },
   *   config: {
   *     [SitemapGenerator.configKey]: {
   *       ...,
   *       events: {
   *         callback: ['document-save', 'document-delete'],
   *         validateConfig: ['validate-config'],
   *       },
   *     },
   *   },
   * };
   * SitemapGenerator.register(context);
   * @static
   */
  static register(context) {
    if (!context || !context.hooks || typeof context.hooks.on !== 'function') {
      throw new Error("Missing event dispatcher in 'context.hooks.on(event, callback)' format.");
    }
    const config = { ...SitemapGenerator.defaultConfig(), ...context.config[SitemapGenerator.configKey] };
    if (!config.events) {
      throw new Error("Missing events to listen to for in 'config.events'.");
    }
    Object.keys(config.events).forEach((method) => {
      config.events[method].forEach((event) => {
        if (typeof SitemapGenerator[method] !== 'function') {
          debug(`Missing function "${method}" for key "${event}"`);
          return;
        }
        context.hooks.on(event, SitemapGenerator[method]);
      });
    });
  }

  /**
   * Wrapper function for calling generating and writing the sitemap file.
   *
   * @static
   * @async
   * @param {object} _document A Uttori document (unused).
   * @param {object} context A Uttori-like context.
   * @param {object} context.config A provided configuration to use.
   * @param {SitemapGeneratorConfig} context.config.configKey - A configuration object specifically for this plugin.
   * @param {object} context.hooks An event system / hook system to use.
   * @param {Function} context.hooks.on An event registration function.
   * @param {Function} context.hooks.fetch An event dispatch function that returns an array of results.
   * @returns {Promise} The provided document.
   * @example <caption>SitemapGenerator.callback(_document, context)</caption>
   * const context = {
   *   config: {
   *     [SitemapGenerator.configKey]: {
   *       ...,
   *     },
   *   },
   *   hooks: {
   *     on: (event) => { ... }
   *   },
   * };
   * SitemapGenerator.callback(null, context);
   */
  static async callback(_document, context) {
    debug('callback');
    const { directory, filename, extension } = { ...SitemapGenerator.defaultConfig(), ...context.config[SitemapGenerator.configKey] };
    const sitemap = await SitemapGenerator.generateSitemap(context);
    debug(`File: ${directory}/${filename}.${extension}`);

    const target = `${directory}/${sanitize(`${filename}.${extension}`)}`;
    try { fs.writeFileSync(target, sitemap, 'utf8'); } catch (error) {
      /* istanbul ignore next */
      debug('Error writing file:', target, sitemap, error);
    }
    return _document;
  }

  /**
   * Generates a sitemap from the provided context.
   *
   * @param {object} context A Uttori-like context.
   * @param {object} context.config A provided configuration to use.
   * @param {SitemapGeneratorConfig} context.config.configKey A configuration object specifically for this plugin.
   * @param {object} context.config.events An object whose keys correspong to methods, and contents are events to listen for.
   * @param {object} context.hooks An event system / hook system to use.
   * @param {Function} context.hooks.on An event registration function.
   * @param {Function} context.hooks.fetch An event dispatch function that returns an array of results.
   * @returns {Promise} The generated sitemap.
   * @example <caption>SitemapGenerator.callback(_document, context)</caption>
   * const context = {
   *   config: {
   *     [SitemapGenerator.configKey]: {
   *       ...,
   *     },
   *   },
   *   hooks: {
   *     on: (event) => { ... },
   *     fetch: (event, query) => { ... },
   *   },
   * };
   * SitemapGenerator.generateSitemap(context);
   * @static
   */
  static async generateSitemap(context) {
    debug('generateSitemap');
    const {
      base_url,
      page_priority,
      url_filters,
      urls,
      xml_footer,
      xml_header,
    } = { ...SitemapGenerator.defaultConfig(), ...context.config[SitemapGenerator.configKey] };

    let documents = [];
    try {
      [documents] = await context.hooks.fetch('storage-query', "SELECT 'slug', 'createDate', 'updateDate' FROM documents WHERE slug != '' ORDER BY updateDate DESC LIMIT 10000");
    } catch (error) {
      /* istanbul ignore next */
      debug('Error geting documents:', error);
    }

    // Add all documents to the urls array
    documents.forEach((document) => {
      urls.push({
        url: `/${document.slug}`,
        lastmod: document.updateDate ? new Date(document.updateDate).toISOString() : new Date(document.createDate).toISOString(),
        priority: page_priority,
      });
    });

    // eslint-disable-next-line unicorn/consistent-function-scoping
    let urlFilter = (input) => input;
    /* istanbul ignore else */
    if (Array.isArray(url_filters) && url_filters.length > 0) {
      urlFilter = (route) => {
        let pass = true;
        url_filters.forEach((url_filter) => {
          try {
            if (url_filter.test(route.url)) {
              pass = false;
            }
          } catch (error) {
            /* istanbul ignore next */
            debug('Sitemap Filter Error:', error, url_filter, route.url);
          }
        });
        return pass;
      };
    }

    const data = urls.reduce((accumulator, route) => {
      if (urlFilter(route)) {
        accumulator += `<url><loc>${base_url}${route.url}</loc>`;
        /* istanbul ignore else */
        if (route.lastmod) {
          accumulator += `<lastmod>${route.lastmod}</lastmod>`;
        }
        /* istanbul ignore else */
        if (route.priority) {
          accumulator += `<priority>${route.priority}</priority>`;
        }
        /* istanbul ignore next */
        if (route.changefreq) {
          accumulator += `<changefreq>${route.changefreq}</changefreq>`;
        }
        accumulator += '</url>';
      }
      return accumulator;
    }, '');

    return `${xml_header}${data}${xml_footer}`;
  }
}

module.exports = SitemapGenerator;
